ssa/normalize.go

```diff
@@ -0,0 +1,161 @@
+	typedObject, err := scheme.New(object.GroupVersionKind())
+	if err != nil {
+		return nil, err
+	}
+
+	metaObject, ok := typedObject.(metav1.Object)
+	if !ok {
+		return nil, err
+	}
+
+	if err = runtime.DefaultUnstructuredConverter.FromUnstructured(object.Object, metaObject); err != nil {
+		return nil, err
+	}
+	return metaObject, nil
+}
+
+// ToUnstructured converts a typed Kubernetes resource into the Unstructured
+// equivalent.
+func ToUnstructured(object metav1.Object) (*unstructured.Unstructured, error) {
+	u, err := runtime.DefaultUnstructuredConverter.ToUnstructured(object)
+	if err != nil {
+		return nil, err
+	}
+	return &unstructured.Unstructured{Object: u}, nil
+}
+
+// NormalizeUnstructured normalizes an Unstructured object by converting it to
+// a typed Kubernetes resource, normalizing it, and then converting it back to
+// an Unstructured object. It only works for API types registered with the
+// default client-go scheme. If the conversion fails, only certain standard
+// fields are removed.
+func NormalizeUnstructured(object *unstructured.Unstructured) error {
+	return NormalizeUnstructuredWithScheme(object, defaultScheme)
+}
+
+// NormalizeUnstructuredWithScheme normalizes an Unstructured object by
+// converting it to a typed Kubernetes resource, normalizing it, and then
+// converting it back to an Unstructured object. It only works for API types
+// registered with the given scheme. If the conversion fails, only certain
+// standard fields are removed.
+func NormalizeUnstructuredWithScheme(object *unstructured.Unstructured, scheme *runtime.Scheme) error {
+	if typedObject, err := FromUnstructuredWithScheme(object, scheme); err == nil {
+		switch o := typedObject.(type) {
+		case *corev1.Pod:
+			normalizePodProtoDefault(&o.Spec)
+		case *appsv1.Deployment:
+			normalizePodProtoDefault(&o.Spec.Template.Spec)
+		case *appsv1.StatefulSet:
+			normalizePodProtoDefault(&o.Spec.Template.Spec)
+		case *appsv1.DaemonSet:
+			normalizePodProtoDefault(&o.Spec.Template.Spec)
+		case *appsv1.ReplicaSet:
+			normalizePodProtoDefault(&o.Spec.Template.Spec)
+		case *batchv1.Job:
+			normalizePodProtoDefault(&o.Spec.Template.Spec)
+		case *batchv1.CronJob:
+			normalizePodProtoDefault(&o.Spec.JobTemplate.Spec.Template.Spec)
+		case *corev1.Service:
+			normalizeServiceProtoDefault(&o.Spec)
+		case *corev1.Secret:
+			normalizeSecret(o)
+		}
+
+		normalizedObject, err := ToUnstructured(typedObject)
+		if err != nil {
+			return err
+		}
+		object.Object = normalizedObject.Object
+	}
+
+	// Ensure the object has an empty creation timestamp, to avoid
+	// issues with the Kubernetes API server rejecting the object
+	// or causing any spurious diffs.
+	unstructured.SetNestedField(object.Object, nil, "metadata", "creationTimestamp")
+
+	// To ensure kstatus continues to work with CRDs, we need to keep the
+	// status field for CRDs.
+	if !IsCRD(object) {
+		unstructured.RemoveNestedField(object.Object, "status")
+	}
```

Comment:
I have my doubts about this actually being required at this point, as we at least seem to do (part of) this again in `manager_diff.go`. If the goal here is to prepare the resource for an apply, wouldn't it be better for this to be done as a preparation step _before_ the apply?

Reply:
Removing it also does not appear to make any of the tests fail, except for the `NormalizeUnstructured` tests I added myself.

Reply:
If we remove the status for all including CRD, then how are we going to add it back before apply?

Reply:
My point is that we would not remove (or modify) anything here at all, we would just normalize the object as-is, and if required, prepare it right before an apply and/or diff.

Due to a lack of test cases covering the scenario in which the existing modification would be required, I can however not be a good judge here.

Reply:
We have such a test, the that reconciles CRDs and CRs

Reply:
It continues to work if I remove L112-121.

Reply:
As this needs much more testing, I will (eventually) do the remainder of this in a separate PR.